source/plugin/entities/ai.rst

```diff
@@ -0,0 +1,195 @@
+take an idle pose (look straight ahead). The AI does not affect any sound effects that an entity might play randomly.
+
+.. note::
+
+    Please note that editing the goals or tasks of an entity will directly edit the entity. This behavior differs from
+    other parts of the API where only copies are returned which must be applied afterwards. Thus it is not possible to
+    transfer the tasks from one entity to another.
+
+.. note::
+    
+    Any changes made to an entities AI will be lost after a server restart or after the entity was unloaded.
+
+Tasks
+-----
+
+The behavior of many entity types is similar to each other. For this reason was the entity AI split into multiple
+smaller and reusable parts called :javadoc:`AITask`\s. Each of the ``AITask``\s represents a single behavior trait of an
+entity such as :javadoc:`AITaskTypes#WATCH_CLOSEST {WATCH_CLOSEST}`, which makes the entity look at the nearest matching
+entity, or :javadoc:`AITask#AVOID_ENTITY {AVOID_ENTITY}`, which makes the entity flee from certain matching entities.
+
+.. note::
+
+    Vanilla Minecraft itself provides a huge variety of ``AITask``\s, however most of them are not yet accessible via
+    the API.
+
+Adding additional AITasks
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Adding additional ``AITask``\s to an ``Entity``\s goal is pretty easy. We start with creating with a simple ``AITask``.
+
+.. code-block:: java
+
+    import org.spongepowered.api.entity.ai.task.builtin.WatchClosestAITask;
+
+    Agent entity = ...;
+    Goal<Agent> goal = ...;
+
+    WatchClosestAITask watchClosestAiTask = WatchClosestAITask.builder()
+            .chance(1)
+            .maxDistance(30)
+            .watch(Player.class)
+            .build(entity);
+    goal.addTask(0, watchClosestAiTask);
+
+In this example we create an :javadoc:`WatchClosestAITask` using the associated builder. Using the build we define that
+the chance of this goal triggering is 100%. Thus whenever a :javadoc:`Player` gets closer to the entity than 30 blocks
+then the entity will look at him. Also we assign zero as a priority, which is a high priority, thus this goal takes
+precedence about almost all other tasks.
+
+.. note::
+
+    Minecraft uses low values as high priority and high values as low priority. By default Minecraft uses priority
+    values from zero to roughly ten.
+
+Removing certain AITasks
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Removing the correct ``AITask``\s can be a little tricky, especially if modded entities or custom AI come into play.
+The Sponge API tries to provides a disambiguator. Calling :javadoc:`AITask#getType()` returns an :javadoc:`AITaskType`
+which can be used to differentiate between the existing types of tasks.
+
+First we try the simple version which will only work if there are no AI altering mods present:
+
+.. code-block:: java
+
+    Goal<Zombie> goal = ...;
+    
+    AITask<Zombie> attackTask = (AITask<Zombie>) goal.getTasks().get(1); // EntityAIZombieAttack
```

Comment:
Should be `AITask<? super Zombie>`

Reply:
What would be the benefit? I still cannot avoid the cast!
The method returns `? super AITask<? extends Zombie>` aka. `Object`, which is almost harassment.

Reply:
No, that method itself is broke (I pointed out before to zidane). The correct type returned should be `<T extends AITask<? super Zombie>>`. For example, if there is a `WalkToTargetAITask extends AITask<Agent>`, `WalkToTargetAITask` should be applicable to a `Zombie`.

Reply:
This is trivial and can be disregarded for now.